### DISCUSSION TITLE
Why is IllegalIdentifierName case insensitive?

### DISCUSSION THREAD

**Discussion by delanym** [checkstyle/checkstyle#12689] [Q&A] [answered]
I want to define an enum describing the states of an http transport as OPEN|CLOSED
With this rule on

``` xml
<module name="IllegalIdentifierName">
  <property name="format" value="(?i)^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$"/>
</module>
```

I can't do it because it doesn't respect case sensitivity. Why is that? Java is a case sensitive language!
There are no options to control case sensitivity documented here: https://checkstyle.sourceforge.io/config_naming.html#IllegalIdentifierName

And here's something confusing: https://github.com/checkstyle/checkstyle/issues/8970

**Comment by nrmancuso** [comment, marked as answer]
> There are no options to control case sensitivity documented here

I think a separate option for this would be overkill, why not remove the case insensitive flag from the regexp?

**Reply by rnveach** [reply]
`(?i)` is the flag telling the regular expression to make this case insensitive. If you wish, you can remove this flag and be case sensitive.

Example:
````
$ cat TestClass.java
public class TestClass {
    void method() {
        String open = "";
        String OPEN = "";
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <property name="haltOnException" value="false"/>

    <module name="TreeWalker">
<module name="IllegalIdentifierName">
  <property name="id" value="IllegalIdentifierName1"/>
  <property name="format" value="(?i)^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$"/>
</module>

<module name="IllegalIdentifierName">
  <property name="id" value="IllegalIdentifierName2"/>
  <property name="format" value="^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$"/>
</module>
    </module>
</module>

$ java -jar checkstyle-10.6.0-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:3:16: Name 'open' must match pattern '(?i)^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$'. [IllegalIdentifierName1]
[WARN] TestClass.java:3:16: Name 'open' must match pattern '^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$'. [IllegalIdentifierName2]
[WARN] TestClass.java:4:16: Name 'OPEN' must match pattern '(?i)^(?!(record|yield|var|permits|sealed|open|transitive|_)$).+$'. [IllegalIdentifierName1]
Audit done.
````

Notice line 4 has only 1 violation, and it is on the case insensitive. Case sensitive ignored the field.

**Reply by romani** [reply]
I have never heard about such option to define case sensitive in regexp :)

We usually have Check property for this. 
Most regexp Checks has such property.
Should we add it?
Or update default property value to be case sensitive and update doc  to point user how to do this.

We have this problem at https://github.com/checkstyle/checkstyle/pull/12665#pullrequestreview-1264885014
Lack of case sensitive make it hard to use, as it raising a lot of false positives

**Reply by nrmancuso** [reply]
> Or update default property value to be case sensitive and update doc to point user how to do this.

This makes more sense to me, no need to duplicate functionality of regexp.

We should probably also update https://checkstyle.sourceforge.io/property_types.html#Pattern to mention some common flags, like `i`.
